scylla/src/transport/connection.rs

```diff
@@ -743,24 +764,43 @@ impl Connection {
         // of the channel will be dropped, this task will return an error
         // and the whole worker will be stopped
         while let Some(task) = task_receiver.recv().await {
-            let stream_id = {
-                // We are guaranteed here that handler_map will not be locked
-                // by anybody else, so we can do try_lock().unwrap()
-                let mut lock = handler_map.try_lock().unwrap();
-
-                if let Some(stream_id) = lock.allocate(task.response_handler) {
-                    stream_id
-                } else {
-                    // TODO: Handle this error better, for now we drop this
-                    // request and return an error to the receiver
-                    error!("Could not allocate stream id");
-                    continue;
+            match task {
+                Task::Request {
+                    stream_id_sender,
+                    response_handler,
+                    serialized_request,
+                } => {
+                    let stream_id = {
+                        let mut hmap = handler_map.try_lock().unwrap();
+                        // TODO: Handle stream id allocation error better,
+                        // for now the error is propagated to the connection
+                        if let Some(id) = hmap.allocate(response_handler) {
+                            stream_id_sender
+                                .send(id)
+                                .map_err(|_| QueryError::UnableToAllocStreamId)?;
+                            id
+                        } else {
+                            error!("Unable to allocate stream id");
+                            return Err(QueryError::UnableToAllocStreamId);
+                        }
+                    };
+                    let mut req = serialized_request;
+                    req.set_stream(stream_id);
+                    write_half.write_all(req.get_data()).await?;
                 }
-            };
-
-            let mut req = task.serialized_request;
-            req.set_stream(stream_id);
-            write_half.write_all(req.get_data()).await?;
+                Task::Orphan { stream_id } => {
+                    let mut hmap = handler_map.try_lock().unwrap();
```

Comment:
Here, we probably can remove the entry from `handler_map` corresponding to the orphaned stream ID.

Reply:
Why? The id is orphaned, but a response to it can still arrive, so once it does, it will remove the entry as well. And if it doesn't and eventually we hit max orphan count, the connection dies anway. Or do you mean that only as an optimization?

Reply:
I mainly meant it as an optimization - the memory used by the entry can be freed at this point because the contents of the entry won't be used (it's a Sender of a oneshot channel which won't be read from).